eva/optimizer/rules/rules.py

```diff
@@ -214,6 +216,39 @@ def apply(self, before: LogicalProject, context: OptimizerContext):
         yield new_get_opr
 
 
+class CacheFunctionExpressionInApply(Rule):
+    def __init__(self):
+        pattern = Pattern(OperatorType.LOGICAL_APPLY_AND_MERGE)
+        pattern.append_child(Pattern(OperatorType.DUMMY))
+        super().__init__(RuleType.CACHE_FUNCTION_EXPRESISON_IN_APPLY, pattern)
+
+    def promise(self):
+        return Promise.CACHE_FUNCTION_EXPRESISON_IN_APPLY
+
+    def check(self, before: LogicalApplyAndMerge, context: OptimizerContext):
+        expr = before.func_expr
+        # already cache enabled
+        # replace the cacheable condition once we have the property supported as part of the UDF itself.
+        if expr.has_cache() or expr.name not in CACHEABLE_UDFS:
+            return False
+        # we do not support caching function expression instances with multiple arguments or nested function expressions
+        if len(expr.children) > 1 or not isinstance(
+            expr.children[0], TupleValueExpression
+        ):
+            return False
+        return True
+
+    def apply(self, before: LogicalApplyAndMerge, context: OptimizerContext):
```

Comment:
I assume we eventually will have a rule that moves function expression from projection and filter into LogicalApplyAndMerge.